In [24]:
import pandas as pd 
from pandasql import sqldf

print("Libraries Loaded Successfully")

Libraries Loaded Successfully


In [25]:
df = pd.read_excel(
    "/workspaces/decodelabs-sql-analysis/dataset/cleaned_dataset.xlsx"
)

df.head()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


In [26]:
df.info()

df.shape

<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   str           
 1   Date             1200 non-null   datetime64[us]
 2   CustomerID       1200 non-null   str           
 3   Product          1200 non-null   str           
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   str           
 7   PaymentMethod    1200 non-null   str           
 8   OrderStatus      1200 non-null   str           
 9   TrackingNumber   1200 non-null   str           
 10  ItemsInCart      1200 non-null   int64         
 11  CouponCode       1200 non-null   str           
 12  ReferralSource   1200 non-null   str           
 13  TotalPrice       1200 non-null   float64       
dtypes: datetime64[us](1), float64(2), int64(2), str(9)


(1200, 14)

In [27]:
pysqldf = lambda q: sqldf(q, globals())

In [28]:
df.columns

Index(['OrderID', 'Date', 'CustomerID', 'Product', 'Quantity', 'UnitPrice',
       'ShippingAddress', 'PaymentMethod', 'OrderStatus', 'TrackingNumber',
       'ItemsInCart', 'CouponCode', 'ReferralSource', 'TotalPrice'],
      dtype='str')

# SQL Data Analysis Project

## Objective

the objective of this project is to analyze -commerce transaction data using SQL queries and derive meaningful business insights.

The analysis focuses on:

- Order Volume
- Revenue Generation
- Customer Behaviour
- Product Performance
- Payment Preferences
- Order Fulfillment Trends

Dataset Sorce: E-Commerce Sales Dataset
Tools Used: Python, Pandas, SqLite Sql (PandaSQL)

## Query 1: Total Number of Orders

### Business Question

How many orders are present in the dataset?

In [29]:
query1 = """
SELECT Count(*) AS Total_Orders
From df
"""

result1 = pysqldf(query1)

result1

,Total_Orders
0,1200


### Insight

The dataset contains 1200 orders, representing the total transaction volume available for analysis. This provides a sufficient sample size to identify customer purchasing patterns, revenue trend, and operational performance metrics.

## Query 2: Revenue Generated by Each Product

### Business Question

Which products generate the highest revenue?

In [30]:
query2 = """
SELECT Product,
    Round(SUM(TotalPrice), 2) AS Revenue
FROM df
GROUP BY Product
ORDER BY Revenue DESC
"""

result2 = pysqldf(query2)

result2

,Product,Revenue
0,Chair,195620.11
1,Printer,195612.61
2,Laptop,192126.56
3,Tablet,186568.95
4,Monitor,175651.41
5,Desk,167459.93
6,Phone,151722.39


### Insight

Chair generated the highest revenue (195,620.11/-), closely followed by Printer (195,612.61/-) and Laptop (192,126.56/-). These products contribute significantly to overall business revenue and should be prioritized for inventory management, promotions, and customer targeting strategies. Phone generated the lowest revenue among the listed products, indicating comparatively lower sales performance.

## Query 3: Average Order Value

### Business Question

What is the average value of an order?

In [31]:
query3 = """
SELECT ROUND(AVG(TotalPrice), 2) AS Average_Order_Value
FROM df
"""

result3 = pysqldf(query3)

result3

,Average_Order_Value
0,1053.97


### Insight

The average order value across all transactions is 1053.97/-. This indicates that customers typically spend around 1000/- per purchase. Increasing average order value through bundle offers, product recommendations, and upselling strategies could significantly improve overall revenue without increasing customer acquisition costs.


Since the average order value is above 1000/-, the business appears to be dealing with medium-to-high value transactions. Marketing campaigns focused on premium and cross-selling opportunities may generate higher return compared to discount-focused strategies.

## Query 4: Orders by Payment Method

### Business Question

Which payment methods are most frequently used by customers?

In [32]:
query4 = """
SELECT
    PaymentMethod,
    COUNT(*) AS Total_Orders
FROM df
GROUP BY PaymentMethod
ORDER BY Total_Orders DESC
"""

result4 = pysqldf(query4)

result4

,PaymentMethod,Total_Orders
0,Online,258
1,Cash,246
2,Credit Card,234
3,Debit Card,232
4,Gift Card,230


### Insight

Online payment is the most frequently used payment method  with 258 orders, followed by Cash (246 orders), Credit Card (234 orders), Debit Card (232 orders), and Gift Cards (230 orders).

The distribution is fairly balanced across all payment methods, indicating that customers prefer multiple payment options rather than relying on single mode of payment.


The company should continue supporting all major payment methods since customer usage is distributed across different options. However, optimizing the online payment experience could further improve customer convenience and checkout efficiency.

## Query 5: Order Status Analysis

### Business Question

What is the distribution of customer order statuses?

In [33]:
query5 = """
SELECT 
    OrderStatus,
    COUNT(*) AS Total_Orders
FROM df
GROUP BY OrderStatus
ORDER BY Total_Orders DESC
"""

result5 = pysqldf(query5)

result5

,OrderStatus,Total_Orders
0,Cancelled,250
1,Returned,247
2,Pending,237
3,Shipped,235
4,Delivered,231


### Insight

Order Statuses are distributed relatively evenly across all categories, including delivered, shipped, pending, returned, and cancelled orders.

This balanced distribution provides a useful dataset for analyzing different stages of the order lifecycle and understanding how order outcomes may impact overall business performance.


The company should continuously monitor order status trends and identify factors contributing to cancellations, returns and pending orders. Reducing unsuccessful order outcomes can improve customer satisfaction and operational efficiency.

## Query 6: Top Customers by Revenue

### Business Question

Which customers generate the highest revenue for the business?


In [34]:
query6 = """
SELECT
    CustomerID,
    ROUND(SUM(TotalPrice),2) AS Revenue
FROM df
GROUP BY CustomerID
ORDER BY Revenue DESC
LIMIT 10
"""

result6 = pysqldf(query6)

result6

,CustomerID,Revenue
0,C38840,5723.23
1,C57276,3456.40
2,C67260,3390.80
3,C13877,3384.90
4,C18404,3370.20
5,C16775,3353.75
6,C65986,3352.40
7,C47778,3334.00
8,C59183,3322.55
9,C25276,3313.90


### Insight

Customer C38840 generated the highest revenue of 5723.23/-, significantly higehr than other customers. The remaining top customers contributed between 3313/- and 3456/- in revenue.

This indicates the presence of high-value customers who contribute substantially to overall business revenue.

The business should focus on retaining high-value customers through loyalty programs, exclusive offers, and personalized marketing campaigns. Customer retention is generally more cost-effective than acquiring new customers.

## Query 7: Revenue by Payment Method

### Business Question

Which payment methods generate the highest revenue?

In [35]:
query7 = """
SELECT 
    PaymentMethod,
    ROUND(SUM(TotalPrice),2) AS Revenue
FROM df
GROUP BY PaymentMethod
ORDER BY Revenue DESC
"""

result7 = pysqldf(query7)

result7

,PaymentMethod,Revenue
0,Credit Card,263847.63
1,Online,262442.94
2,Cash,259786.29
3,Gift Card,246323.92
4,Debit Card,232361.18


### Insight

Credit Card payments generated the highest revenue (263,847.63/-), despite having fewer transactions than Online payments.

This suggests that customers using Credit Cards tend to place higher-value orders compared to customers using other payment methods.


The company should consider offering premium benefits, reward programs, or targeted promotions for credit card users, as they represent a high-value customer segment that contributes significantly to revenue.

## Query 8: Products by Quantity Sold

### Business Question

Which products are sold most frequently?

In [36]:
query8 = """
SELECT
    Product,
    SUM(Quantity) AS Total_Quantity_Sold
FROM df
GROUP BY Product
ORDER BY Total_Quantity_Sold DESC
"""

result8 = pysqldf(query8)

result8

,Product,Total_Quantity_Sold
0,Chair,562
1,Printer,542
2,Laptop,535
3,Desk,508
4,Tablet,497
5,Monitor,480
6,Phone,411


### Insight

Chair recorded the highest quantity sold (562 units), followed by Printer (542 units) and Laptop (535 units). These products demomstrate strong customer demand and contribute significantly to sales volume.

Phone recorded the lowest quantity sold (411 units), indicating comparatively lower demand.


The company should prioritize inventory availability for high-demand products such as Chairs, Printers and Laptops to prevent stock shortages. Marketing campaigns and promotional offers can be used to improve sales of lower-performing products.

## Query 9: Revenue by Order Status

### Business Question

Which order statuses contribute the most revenue?

In [37]:
query9 = """
SELECT 
    OrderStatus,
    ROUND(SUM(TotalPrice),2) AS Revenue
FROM df
GROUP BY OrderStatus
ORDER BY Revenue DESC
"""

result9 = pysqldf(query9)

result9

,OrderStatus,Revenue
0,Cancelled,276396.21
1,Pending,256328.15
2,Shipped,246159.58
3,Returned,243277.70
4,Delivered,242600.32


### Insight

Revenue is distributed relatively evenly across all order statuses. Cancelled orders generated the highest recorded revenue (276,396.21/-), followed by Pending (256,328.15/-), Shipped (246,159.58/-), Returned (243,277.70/-), and Delivered orders (242,600.32/-).

The balanced distributed indicates that the dataset contains diverse transaction scenarios, making it suitable for analytical and SQL practice.

Organizations should closely monitor order lifecycle stages and implement strategies to convert pending orders into completed deliveries while minimizing cancellations and returns. Improved order fulfillment processes can positively impact revenue generation and customer satisfaction.

## Query 10: Referral Source Performance

### Business Question

Which referral sources bring the highest number of customers?

In [38]:
query10 = """
SELECT
    ReferralSource,
    COUNT(*) AS Total_Customers
FROM df
GROUP BY ReferralSource
ORDER BY Total_Customers DESC
"""

result10 = pysqldf(query10)

result10

,ReferralSource,Total_Customers
0,Instagram,259
1,Email,250
2,Google,241
3,Facebook,228
4,Referral,222


### Insight

Instagram generated the highest number of customer acquisitions (259 customers), followed by Email (250) and Google (241).

This indicates that social media and digital marketing channels play a significant role in attracting customers and driving business growth.


The company should continue investing in high-performing acquisition channels such as instagram and Email marketing while optimizing lower-performing channels to improve customer reach and marketing efficiency.

# Summary

This SQL-based business analysis project examined an ecommerce dataset containing customer orders, product information, payment methods, referral sources, and revenue metrics.

Using SQL queries, key business insights were extracted regarding revenue generation, customer purchasing behaviour, product performance, payment preferences, order lifecycle status, and marketing channel effectiveness.

The analysis revealed that:

- Chair and Printer products generated the highest sales performance.
- Credit Card transactions generated the highest revenue.
- The average order value was 1053.97/-
- Certain customers contributed significantly more revenue than others.
- Instagram was the leading customer acquisition channel.
- Product demand varied across categories, creating opprotunities for inventory optimization.

The findings can help businesses improve decision-making related to marketing, customer retention, inventory management, and revenue optimization.